$a)$ Estimar el numero de intervalo para aproximar $\int_a^b f(x)dx$ con una tolerancia de $10^{-5}$

In [2]:
from sympy import symbols,integrate, log, exp, sin, diff
import numpy as np

# Regla del Trapecio compuesto
Dividimos el intervalo $[a,b]$ en pequeños subintervalos (n par) de manera que
$$x_I = a+ih \text{ donde } h=\frac{b-a}{n} \text{ para } i=0,1,2...,n$$
Aplicando regla se simpson a cada subintervalo $[x_{j-1},x_{i+1}], j=1,3,5,...,n-1$ tenemos
 $$\int_{x_{j-1}}^{x_{j+1}}f(x) \approx \frac{x_{j+1}-x_{j}}{2} [f(x_{j})+f(x_{j+1})] $$
 Sumando las integrales de todos los subintervalos:
 $$\int_a^b f(x)dx \approx \frac{h}{2}  \left [ \sum_{j=0}^{n-1} f(x_{j})+ f(x_{j+1})\right ]$$

##Ejemplo:
consideremos $f(x) = \sqrt{4 + 5sen^2(t)}$, hallemos
$$\int_0^{\pi/4} f(x)dx= \int_0^{\pi/4} \sqrt{4 + 5sen^2(t)}\; dt$$

In [20]:
def Trapecio(a,b,n,f):
  h = (b-a)/n
  x0 = [a+i*h for i in range(n+1)]
  suma = sum(f.subs(x,x0[j]) + f.subs(x,x0[j+1]) for j in range(n))
  return h*suma/2

In [21]:
x = symbols('x')
f2 = (4+5*(sin(x))**2)**0.5
# Calculo del error
diff(f2, x, 2)

-25.0*sin(x)**2*cos(x)**2/(5*sin(x)**2 + 4)**1.5 - 5.0*sin(x)**2/(5*sin(x)**2 + 4)**0.5 + 5.0*cos(x)**2/(5*sin(x)**2 + 4)**0.5

\begin{align}
|f^{(2)}(\xi)| &\leq \frac{0}{4^{1.5}} + \frac{0}{4^{0.5}} + \frac{5.0\cdot 1}{4^{0.5}}\\
& = 2.5
\end{align}

El error entonces:
\begin{align}
|E(x)| &=\frac{h^3}{12}n \cdot |f^{(2)}(\xi)|\\
& \leq \frac{h^3}{12}n\cdot 2.5
\end{align}
como la tolerancia es de $10^{-5}$

\begin{align}
 &\frac{h^3}{12}n \cdot 2.5 < 10^{-5} \\
\rightarrow\; & h^3n < 0.000048  
\end{align}
Asi $n>  100.46486 $\
La integral usando $n=101$ es:

In [22]:
Trapecio(0,np.pi/4,101,f2)

1.73484840290702

In [16]:
inte = integrate(f2, (x,0, np.pi/4)).round(13)
inte

1.7348434616135

In [24]:
print("Iteracion\t Error")
for i in range(1,102,4):
  inte = integrate(f2, (x,0, np.pi/4)).round(13)
  error = inte - Trapecio(0,np.pi/4,i,f2)
  print(i, "\t", abs(error))

Iteracion	 Error
1 	 0.0517448420598299
5 	 0.00201918699182246
9 	 0.000622576631938321
13 	 0.000298324613812184
17 	 0.000174437134118577
21 	 0.000114308693514698
25 	 8.06542456972537e-5
29 	 5.99383427268840e-5
33 	 4.62880139506794e-5
37 	 3.68205267058386e-5
41 	 2.99863481236784e-5
45 	 2.48922789856110e-5
49 	 2.09940534825481e-5
53 	 1.79446817059059e-5
57 	 1.55144727744805e-5
61 	 1.35464794603912e-5
65 	 1.19305067391995e-5
69 	 1.05873433826620e-5
73 	 9.45886668457518e-6
77 	 8.50164708032786e-6
81 	 7.68270623074940e-6
85 	 6.97663802595727e-6
89 	 6.36361410122355e-6
93 	 5.82797637638244e-6
97 	 5.35722716987763e-6
101 	 4.94129352057726e-6


# Regla se Simpson Compuesta
Dividimos el intervalo $[a,b]$ en pequeños subintervalos (n par) de manera que
$$x_I = a+ih \text{ donde } h=\frac{b-a}{n} \text{ para } i=0,1,2...,n$$
Aplicando regla se simpson a cada subintervalo $[x_{j-1},x_{i+1}], j=1,3,5,...,n-1$ tenemos
 $$\int_{x_{j-1}}^{x_{j+1}} \approx \frac{x_{j+1}+x_{j-1}}{6} [f(x_{j-1})+ 4f(x_j)+ f(x_{j+1})] $$
 Sumando las integrales de todos los subintervalos:
 $$\int_a^b f(x)dx \approx \frac{h}{3}  \left [ f(x_0)+ 2\sum_{j=1}^{n/2-1} f(x_{2j}) + 4\sum_{j=1}^{n/2} f(x_{2j-1})+ f(x_n) \right ]$$

In [5]:
def Simpson(a,b,n,f):
  h = (b-a)/n
  x0 = [a+i*h for i in range(n+1)]
  suma = sum(2*f.subs(x,x0[2*j]) + 4*f.subs(x,x0[2*j-1]) for j in range(1,int(n/2)) )
  suma = f.subs(x,x0[0]) + suma + 4*f.subs(x,x0[n-1]) + f.subs(x,x0[n])
  return h*suma/3

In [6]:
x = symbols('x')
f2 = (4+5*(sin(x))**2)**0.5

In [ ]:
integrate(f2, (x,0, 2*np.pi)).round(13)

15.8654395892906

In [7]:
diff(f2, x, 4)

-9375.0*sin(x)**4*cos(x)**4/(5*sin(x)**2 + 4)**3.5 - 2250.0*sin(x)**4*cos(x)**2/(5*sin(x)**2 + 4)**2.5 + 2250.0*sin(x)**2*cos(x)**4/(5*sin(x)**2 + 4)**2.5 - 75.0*sin(x)**4/(5*sin(x)**2 + 4)**1.5 + 550.0*sin(x)**2*cos(x)**2/(5*sin(x)**2 + 4)**1.5 - 75.0*cos(x)**4/(5*sin(x)**2 + 4)**1.5 + 20.0*sin(x)**2/(5*sin(x)**2 + 4)**0.5 - 20.0*cos(x)**2/(5*sin(x)**2 + 4)**0.5

\begin{align}
|f^{(4)}(\xi)| &\leq 0 + 0 + \frac{2250\cdot 4}{27\cdot 4^{2.5}} + 0 + \frac{550\cdot 1}{4\cdot 4^{1.5}} + 0 + \frac{20\cdot 1}{4^{0.5}} + 0 \\
& = 37.60416
\end{align}

El error entonces:
\begin{align}
|E(x)| &=\frac{h^5}{90} \cdot |f^{(4)}(\xi)|\\
&\leq  \frac{h^5}{90}\cdot 37.60416
\end{align}
como tol=$10^-5$
\begin{align}
\frac{h^5}{90}&\cdot 37.60416 < 10^{-5}\\
\rightarrow & h^5 < 6.36459\cdot 10^{-8}
\end{align}
Asi $n > 83.6331 $\
La integral usando $n=84$ (par) es:

In [10]:
Simpson(0,np.pi/4,84,f2)

1.73484346146540

In [19]:
print("Iteracion\t Error")
inte = integrate(f2, (x,0, np.pi/4)).round(13)
for i in range(4,85,4):
  error = inte - Simpson(0,np.pi/4,i,f2)
  print(i, "\t", error)

Iteracion	 Error
4 	 2.86874312853680e-5
8 	 1.79983786119386e-6
12 	 3.55515870076317e-7
16 	 1.12485068592960e-7
20 	 4.60733404761271e-8
24 	 2.22188685228275e-8
28 	 1.19931546915808e-8
32 	 7.03016223013719e-9
36 	 4.38890146448045e-9
40 	 2.87956791744648e-9
44 	 1.96679406094802e-9
48 	 1.38869871157965e-9
52 	 1.00824060211835e-9
56 	 7.49602824257067e-10
60 	 5.68834090941550e-10
64 	 4.39420277942304e-10
68 	 3.44806183605328e-10
72 	 2.74343214812234e-10
76 	 2.20995888255970e-10
80 	 1.80010006900488e-10
84 	 1.48102197172761e-10
